File: app/src/test/java/com/pomo/sync/recovery/RecoveryArtifactsTest.kt
package com.pomo.sync.recovery

import com.pomo.sync.protocol.CborValue
import com.pomo.sync.protocol.DeterministicCbor
import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertTrue
import org.junit.Test

public class RecoveryArtifactsTest {
    @Test
    public fun artifactsAreEncryptedBoundedAndCorruptionFailsClosed() {
        val file = RecoveryArtifactCodec.sealFile(
            RecoveryFileBody(byteArrayOf(1, 2), 4, listOf("a@3"), listOf("cap"), listOf("mail")),
            "correct horse",
        )
        assertTrue(RecoveryArtifactCodec.open(file, "correct horse").isNotEmpty())
        val corrupt = file.copy(
            ciphertextAndTag = file.ciphertextAndTag.copyOf().also { it[0] = (it[0].toInt() xor 1).toByte() },
        )
        assertTrue(runCatching { RecoveryArtifactCodec.open(corrupt, "correct horse") }.isFailure)
        val archiveBytes = DeterministicCbor.encode(
            CborValue.Array(listOf(CborValue.Integer(1), CborValue.Text("manifest"))),
        )
        assertArrayEquals(
            archiveBytes,
            RecoveryArtifactCodec.open(RecoveryArtifactCodec.sealArchive(archiveBytes, "correct horse"), "correct horse"),
        )
    }

    @Test
    public fun archiveAuthorityAndForwardRestoreFailClosed() {
        val manifest = RecoveryArchiveManifest(
            "archive",
            2,
            listOf("checkpoint"),
            listOf("pack"),
            listOf("blob"),
            ByteArray(32),
            null,
        )
        RecoveryArtifactCodec.validateArchiveManifest(manifest, false)
        assertTrue(runCatching { RecoveryArtifactCodec.validateArchiveManifest(manifest, true) }.isFailure)
        val plan = prepareForwardRestore(
            "safety",
            listOf(RestoreSelection(RestoreDomain.HISTORY, "session", byteArrayOf(1))),
        )
        assertTrue(
            runCatching { authorizeForwardRestore(plan, RestorePlanOrigin.AGENT, false, RestoreApproval(true, false)) }.isFailure,
        )
        assertTrue(
            runCatching {
                prepareForwardRestore("safety", listOf(RestoreSelection(RestoreDomain.ACTIVE_PHASE, "phase", byteArrayOf(1))))
            }.isFailure,
        )
    }

    @Test
    public fun historyUsesDomainCausalityNotWallClockRows() {
        val older = HistoricalValue(RestoreDomain.PROFILE, "me", 1, "op-z", null)
        val newer = older.copy(causalVersion = 2, operationId = "op-a")
        assertTrue(compareHistoricalValues(older, newer) < 0)
    }
}

File: extension/tests/sync/recoveryArtifacts.test.ts
import { describe, expect, test } from "bun:test";
import { authorizeForwardRestore, compareHistoricalValues, openRecoveryArtifact, prepareForwardRestore, sealRecoveryArchive, sealRecoveryFile, validateRecoveryArchiveManifest } from "../../src/sync/recovery/recoveryArtifacts";
import { encodeCanonicalCbor } from "../../src/sync/protocol/cbor";

describe("Recovery artifacts and forward restore", () => {
  test("encrypts bounded files and archives and rejects corruption", async () => {
    const file = await sealRecoveryFile({ protectedAuthority: new Uint8Array([1, 2]), recoveryGeneration: 4, frontierEvidence: ["a@3"], capabilityLocators: ["cap"], mailboxLocators: ["mail"] }, "correct horse");
    expect((await openRecoveryArtifact(file, "correct horse")).length).toBeGreaterThan(0);
    const corrupt = { ...file, ciphertextAndTag: file.ciphertextAndTag.slice() }; corrupt.ciphertextAndTag[0]! ^= 1;
    await expect(openRecoveryArtifact(corrupt, "correct horse")).rejects.toThrow();
    const archive = encodeCanonicalCbor([1, "manifest"]); expect(await openRecoveryArtifact(await sealRecoveryArchive(archive, "correct horse"), "correct horse")).toEqual(archive);
  });

  test("archive authority and restore confirmations fail closed", () => {
    const manifest = { archiveId: "archive", recoveryGeneration: 2, checkpointIds: ["checkpoint"], packIds: ["pack"], blobIds: ["blob"], manifestDigest: new Uint8Array(32), credentialProof: null };
    validateRecoveryArchiveManifest(manifest, false);
    expect(() => validateRecoveryArchiveManifest(manifest, true)).toThrow(/credentials/);
    const plan = prepareForwardRestore("safety", [{ domain: "HISTORY", targetId: "session", compensatingPayload: new Uint8Array([1]) }]);
    expect(() => authorizeForwardRestore(plan, "AGENT", false, { humanConfirmed: true, independentConfirmed: false })).toThrow(/independent/);
    expect(() => prepareForwardRestore("safety", [{ domain: "ACTIVE_PHASE", targetId: "phase", compensatingPayload: new Uint8Array([1]) }])).toThrow(/rewind/);
  });

  test("compares domain causality, never row timestamps", () => {
    const older = { domain: "PROFILE" as const, targetId: "me", causalVersion: 1, operationId: "op-z", value: null };
    expect(compareHistoricalValues(older, { ...older, causalVersion: 2, operationId: "op-a" })).toBeLessThan(0);
  });
});
